### PR TITLE
refactor(k8s-plugin): update to use annotation based label selector

### PR DESCRIPTION
**Pull request by moustafab**
## Hey, I just made a Pull Request!

completes work described in https://github.com/backstage/backstage/pull/2977

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by moustafab** [comment]
@mclarke47 @freben I tried to get things working but it seems there is a conflicting change from another module preventing the loading of the page. The message I am seeing is this:
```
Uncaught Error: Missing required config value at 'kubernetes.clusters'
    getConfigArray reader.ts:146
    KubernetesContent KubernetesContent.tsx:112
    renderWithHooks react-dom.development.js:14826
    mountIndeterminateComponent react-dom.development.js:17506
    beginWork react-dom.development.js:18630
    callCallback react-dom.development.js:189
    invokeGuardedCallbackDev react-dom.development.js:238
    invokeGuardedCallback react-dom.development.js:293
    beginWork$1 react-dom.development.js:23235
    performUnitOfWork react-dom.development.js:22189
    workLoopSync react-dom.development.js:22162
    performSyncWorkOnRoot react-dom.development.js:21788
    flushSyncCallbackQueueImpl react-dom.development.js:11112
    unstable_runWithPriority scheduler.development.js:653
    runWithPriority$1 react-dom.development.js:11062
    flushSyncCallbackQueueImpl react-dom.development.js:11107
    flushSyncCallbackQueue react-dom.development.js:11095
    discreteUpdates$1 react-dom.development.js:21925
    discreteUpdates react-dom.development.js:1072
    dispatchDiscreteEvent react-dom.development.js:4169
reader.ts:146
The above error occurred in the <KubernetesContent> component:
    in KubernetesContent (at Router.tsx:39)
    in Route (at Router.tsx:37)
    in Routes (at Router.tsx:36)
    in Router (at EntityPage.tsx:197)
    in article (at Content.tsx:54)
    in Content (at Tabbed.tsx:118)
    in Layout (at EntityPageLayout.tsx:108)
    in div (at Page.tsx:45)
    in ThemeProvider (at Page.tsx:39)
    in Page (at EntityPageLayout.tsx:85)
    in EntityPageLayout (at EntityPage.tsx:168)
    in ServiceEntityPage (at EntityPage.tsx:275)
    in EntityPage (at Router.tsx:57)
    in EntityPageSwitch (at Router.tsx:83)
    in EntityProvider (at Router.tsx:82)
    in Route (at Router.tsx:79)
    in Routes (at Router.tsx:77)
    in Router (at App.tsx:72)
    in Route (at App.tsx:70)
    in Routes (at App.tsx:68)
    in AppRoutes (at App.tsx:96)
    in div (at Page.tsx:66)
    in SidebarPage (at Root.tsx:77)
    in Root (at App.tsx:95)
    in Route (at App.tsx:278)
    in Routes (at App.tsx:277)
    in SignInPageWrapper (at App.tsx:276)
    in Router (created by BrowserRouter)
    in BrowserRouter (at App.tsx:275)
    in AppRouter (at App.tsx:94)
    in CssBaseline (created by WithStyles(CssBaseline))
    in WithStyles(CssBaseline) (at AppThemeProvider.tsx:94)
    in ThemeProvider (at AppThemeProvider.tsx:93)
    in AppThemeProvider (at App.tsx:215)
    in AppContextProvider (at App.tsx:214)
    in ApiProvider (at App.tsx:92)
    in Provider (at App.tsx:91)
    in App (created by HotExportedApp)
    in AppContainer (created by HotExportedApp)
    in HotExportedApp (at src/index.tsx:23)

React will try to recreate this component tree from scratch using the error boundary you provided, AppContainer.```

The config seems correct in my app-config.local.yaml as follows:
```
kubernetes:
  serviceLocatorMethod: 'multiTenant'
  clusterLocatorMethods:
    - 'config'
  clusters:
    - url: https://kubernetes.docker.internal:6443
      name: minikube
      serviceAccountToken: 'eyJhbGciOiJSUzI1NiIsImtpZCI6IlRqNkNy...'
      authProvider: 'serviceAccount'
```

If I completely remove the section I get a different response altogether. Not really sure what is wrong, but when I tried on master it was also broken.

File: plugins/kubernetes/README.md [hunk 1]
```diff
@@ -14,26 +14,10 @@ It is only meant for local development, and the setup for it can be found inside
 
 ## Surfacing your Kubernetes components as part of an entity
 
-There are two ways to surface your kubernetes components as part of an entity.
+There are 2 ways to surface your kubernetes components as part of an entity.
```
Comment (freben) [inline plugins/kubernetes/README.md:17]: why though :D haha, commonly you use the word form for small numbers

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts [hunk 2]
```diff
@@ -50,11 +50,13 @@ export class KubernetesBackendClient implements KubernetesApi {
...
-    requestBody: KubernetesRequestBody,
-  ): Promise<ObjectsByEntityResponse> {
+  async getObjectsByLabelSelector(
+    serviceId: String,
+    labelSelector: String,
+    requestBody: AuthRequestBody,
+  ): Promise<ObjectsByLabelSelectorResponse> {
     return await this.getRequired(
-      `/services/${requestBody.entity.metadata.name}`,
+      `/services/${serviceId}?labelSelector=${labelSelector}`,
```
Comment (freben) [inline plugins/kubernetes/src/api/KubernetesBackendClient.ts:59]: should this one be URI component encoded? maybe it's not even supported

**Review by freben — APPROVED** [review]
Small questions, but ready to go imo

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts [hunk 3]
```diff
@@ -50,11 +50,13 @@ export class KubernetesBackendClient implements KubernetesApi {
     return await response.json();
   }
 
-  async getObjectsByEntity(
-    requestBody: KubernetesRequestBody,
-  ): Promise<ObjectsByEntityResponse> {
+  async getObjectsByLabelSelector(
```
Comment (Rugvip) [inline plugins/kubernetes/src/api/KubernetesBackendClient.ts:53]: Can we keep `getObjectByEntity` and the existing route for querying by entity? We will want more ways to filter objects by entity in the future so this gets a bit too specific

File: plugins/kubernetes/README.md (hunk 1 above)
Comment (moustafab) [reply plugins/kubernetes/README.md:17]: haha yea was pulling in changes from my original approach, not purposely changing that back to a number

**Comment by freben** [comment]
Hey, let's get the `getObjectByEntity` structure back and then we're happy to merge

**Comment by freben** [comment]
gentle nudge @moustafab :)

**Comment by moustafab** [comment]
@freben @Rugvip I think this is probably ready to go. Tested and that issue is gone.

File: plugins/kubernetes-backend/src/service/getKubernetesObjectsForServiceHandler.ts [hunk 4]
```diff
@@ -92,7 +79,10 @@ export const handleGetKubernetesObjectsForService: GetKubernetesObjectsForServic
       .join(', ')}]`,
   );
 
-  const labelSelector = parseLabelSelector(requestBody.entity);
+  const labelSelector: string =
+    requestBody.entity?.metadata?.annotations?.[
+      'backstage.io/kubernetes-label-selector'
+    ] || `backstage.io/kubernetes-id=${requestBody.entity.metadata.name}`;
```
Comment (freben) [inline plugins/kubernetes-backend/src/service/getKubernetesObjectsForServiceHandler.ts:85]: Just noting that the line above is very defensive with `entity?.metadata?.` but this one just gets it plain. Gonna assume that the one on this line is ok.

**Review by freben — APPROVED** [review]
Let's get this in!

**Pull request merged**